DAOS-5919 control: Fix -DER_OOG on pool create

mjmac opened the pull request:
The work landed for DAOS-5385 introduced a small race window
where a server could join and not be included in the CaRT
primary group. This commit resurrects part of the old mgmt svc
implementation to allow for synchronous group updates and async
map broadcast.

daosbuild1 commented on src/mgmt/srv_system.c:
```diff
@@ -0,0 +1,417 @@
+ds_mgmt_group_update_handler(struct mgmt_grp_up_in *in)
+{
+	struct mgmt_svc		*svc;
+	struct server_entry 	*map_servers;
```
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	struct server_entry	*map_servers;
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3743/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/mgmt/SConscript:14](https://github.com/daos-stack/daos/blob/ac21c5be9e4f0fa5d3b13c8cac69932eb3212f52/src/mgmt/SConscript#L14):
(pylint-invalid-name) Invalid variable name "pb"

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/1/execution/node/60/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/5/execution/node/297/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/5/execution/node/259/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/5/execution/node/251/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/6/execution/node/291/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/6/execution/node/274/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3743/6/execution/node/302/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mjmac commented:
@kccain, @liw: Do you see a better way to accomplish this goal? It is admittedly a bit of a kludge to use rsvc this way, but I think trying to drive the map broadcast from the control plane will continue to cause problems. I thought about implementing a simpler dedicated solution instead of repurposing the mgmt svc, but this way we can take advantage of the existing well-tested code.

tanabarr commented on src/control/server/instance.go:
```diff
@@ -230,21 +224,38 @@ func (srv *IOServerInstance) determineRank(ctx context.Context, ready *srvpb.Not
 
+	// hack -- If we receive a nil rank from determineRank() it means
+	// that we can skip the rest of setup because it already happened.
+	if r.Equals(system.NilRank) {
```
would prefer to just return Boolean from determineRank indicating a local join rather than assuming from a NilRank

mjmac replied:
Yeah that's fair. When I added that I wasn't sure yet if this was going to be a short-term or long-term solution. I think I'm pretty happy with it as a long-term solution though so it's worth doing correctly.

tanabarr reviewed (approved):
the go changes looks okay, but we now seem to have a hybrid solution with parts of the MS in the io server. is this a temporary solution ?

mjmac commented:
> 
> 
> the go changes looks okay, but we now seem to have a hybrid solution with parts of the MS in the io server. is this a temporary solution ?

We need to have parts of the MS in the ioserver because the control plane has no way to interact directly with CaRT. The problem that this PR is solving is that we have to be sure that all joined server ranks are added to the CaRT primary group on the ioserver rank that is used for pool management CoRPCs. With the previous implementation, the group update was not synchronous with join, and there were potential races where a rank could join and not be included in the group update. Rather than adding more complex locking which may or may not have solved the problem, I refactored the code such that the group update happens in the same goroutine as the join processing, ensuring that everything is synchronized.

We can't make the map broadcast synchronous, however, due to the fact that the ranks won't have started yet. In an earlier prototype, I had added a separate dRPC call to allow the control plane to asynchronously drive the map broadcast, but decided against that approach because one of the motivations for this work was to reduce contention for the dRPC connection, and a map broadcast can run for a while (particularly as the number of ranks increases). As the rsvc code includes support for asynchronous map broadcast (currently used by the pool service and formerly used by the old MS implementation), it seemed best to use that code.

I'm open to the idea of some refactoring/renaming to make it clear that ioserver component is only a worker for the control plane and is not a secondary source of truth about the system. Another possibility is that we factor the map broadcast out of the rsvc implementation and simplify the worker component so that it's not a pseudo-rsvc implementation.

All that having been said, unless there are serious concerns with the PR, I would prefer to land this as-is in order to fix the intermittent failures that are occurring and then address any refactoring feedback at a more leisurely pace.

mjmac commented:
As an aside, another benefit of this approach is that I think we can remove -DER_OOG as a retryable pool creation error, because the group update is guaranteed to complete before the join responses are sent, meaning that the test harness can't race with the group update when creating a pool now.

kccain reviewed (commented):
just some preliminary discussion on the ioserver part of the change.

mjmac commented on src/mgmt/srv_system.c:
```diff
@@ -0,0 +1,417 @@
+	int			rc;
+
+	/* ensure that it's started */
+	rc = ds_mgmt_svc_start();
```
I think there is a case to be made for adding new dRPC handlers for leadership gain/loss. I'm working on a followup patch to actually add support for multiple MS replicas now, will take a look at this question as part of that work. Bigger-picture, though, I think what I'd like to get some thoughts on is what a non-kludgy solution to the problem might look like. I'm reasonably confident in what I have now as a short-term solution for the immediate problem, but I'm not exactly proud of it.

kjacque reviewed (approved):
This seems reasonable, though I don't have enough insight on the IO server side to know if there's a better way. I would be in favor of renaming on the IO server side to indicate that it's a worker service, and providing some in-repo documentation of how these two pieces work together. I could follow the flow well enough, but I think for the future we'd benefit from a few diagrams/flowcharts.

mjmac commented:
Thanks all for the reviews. I'm going to land this now in order to get it in before the weekly test run. Will address the cleanup of `determineRank()` and removal of `-DER_OOG` as a retryable pool create error in a followup.

liw commented:
> @kccain, @liw: Do you see a better way to accomplish this goal? It is admittedly a bit of a kludge to use rsvc this way, but I think trying to drive the map broadcast from the control plane will continue to cause problems. I thought about implementing a simpler dedicated solution instead of repurposing the mgmt svc, but this way we can take advantage of the existing well-tested code.

Updating the MS leader's local crt group synchronously is certainly welcome. As to asynchronous system map broadcast, I feel we need to start thinking about improving dRPC performance (e.g., supporting concurrent dRPCs), as there will likely be more operations affected.
